Explore: Align Explore with Dashboards and Panels

hugohaggmark opened the pull request:
**What this PR does / why we need it**:
To make Explore more aligned with Grafana we've decided to remove queryTransactions from Explore and just query the datasource with one transaction with all the targets instead.

This is one of several PR that tackle this area, will be followed up with #16808 as well.

The end goal is to have one result in state instead of graphResult, tableResult and logsResult. As a preparation for this the QueryStatus component shows status and latency for the current query batch and not per ResultType.

**Which issue(s) this PR fixes**:
Closes #16819

**Special notes for your reviewer**:

hugohaggmark committed:
Wip: Removes queryTransactions from state

hugohaggmark committed:
Refactor: Adds back query failures

hugohaggmark committed:
Refactor: Moves error parsing to datasources

hugohaggmark committed:
Refactor: Adds back hinting for Prometheus

hugohaggmark committed:
Refactor: removed commented out code

hugohaggmark committed:
Refactor: Adds back QueryStatus

hugohaggmark committed:
Refactor: Adds scanning back to Explore

hugohaggmark committed:
Fix: Fixes prettier error

torkelo reviewed (commented):
Excited that we are making progress on this front! 

This looks like a good start. Looks like you could need some help testing this, have only looked at the code so far, will try to find time to test some as well.

hugohaggmark committed:
Fix: Makes sure there is an error

marefr reviewed (commented):
Done some testing and not looked into details of the code yet. So far so good. 
* Prometheus hints seems to work, both rate and histogram :+1: 
* Switching from prometheus query ?`go_gc_duration_seconds{job="grafana"}` to Loki keeps the `{job="grafana"} :+1: 
* Loki scanning works as expected :+1: 
* Using test datasource works as expected with graph and table result :+1: 
* Have seen some weird behavior when splitting, but haven't been able to reproduce yet.
* Thought I found an issue with Prometheus first, but regression in master. Opened PR #16829

marefr commented on public/app/core/utils/explore.ts:
```diff
@@ -184,6 +181,20 @@ export const safeParseJson = (text: string) => {
   }
 };
 
+export const safeStringifyValue = (value: any, space?: number) => {
```
This doesn't seem to be used?

hugohaggmark replied:
Good catch, it should be used when parsing errors. I'll take a look.

marefr commented on public/app/core/utils/explore.ts:
```diff
@@ -280,27 +306,19 @@ export function hasNonEmptyQuery<TQuery extends DataQuery = any>(queries: TQuery
+  const tableResult =
+    resultType === 'Table' && result && result.columns && result.rows
+      ? mergeTablesIntoModel(new TableModel(), result)
+      : mergeTablesIntoModel(new TableModel());
```
This doesn't seem to work. Tried with using test datasource with a Random Walk and Table Static, but only the Graph are populated. However two legends are rendered for the graph which is wrong. Probably because result can include both series and tables now compared to before. 

This is something that will be easier handled in later refactorings so we just need to make sure that this works for now.

Note that for test datasource to even show a table in explore you need to change plugin.json so that `"tables": true,` and rebuild backend.

ryantxu replied:
Hopefully any issue with table vs timeseries should be solved by using SeriesData :)

hugohaggmark replied:
Very good catch, I've fixed it. See if you can repro it again.

marefr reviewed (changes requested):
Had a look at the code and found some issues, see comments. Think it looks great. Much easier to reason about the actions query started/success/failure than transaction started :+1: 

Need to merge with master to get the changes from #16793

hugohaggmark committed:
Merge: Merges with master

hugohaggmark committed:
Fix: Adds safeStringifyValue to error parsing

hugohaggmark committed:
Fix: Fixes table result calculations

hugohaggmark committed:
Refactor: Adds ErrorContainer and generic error handling in Explore

hugohaggmark commented:
@marefr when you have time please take another look at the changes for the error handling.

hugohaggmark committed:
Fix: Fixes so refIds remain consistent

marefr reviewed (changes requested):
RefId's now starts at A - yay :+1: 

Regarding error handling it generally works as expected:
- Query errors for Loki and Prometheus are displayed below related query row :+1: 
- However, if you have multiple queries for Loki and Prometheus there's no result rendered in graph/table/log

![image](https://user-images.githubusercontent.com/1668778/57374667-a98d7380-719b-11e9-8176-7ab244f84a0e.png)

Side note: Switching to Prometheus from above query error and the error is still visible, see screenshot below:
![image](https://user-images.githubusercontent.com/1668778/57374837-202a7100-719c-11e9-8eae-98ea518c714b.png)

![image](https://user-images.githubusercontent.com/1668778/57374861-320c1400-719c-11e9-8f64-21ebaec2e354.png)

Testing v6.2-beta1 on play.grafana.org there's result rendered in graph/table/log even though one of the queries fails:

![image](https://user-images.githubusercontent.com/1668778/57374913-536d0000-719c-11e9-97af-db475be23460.png)

Testing error handling of datasource that not supports showing error below query field, elasticsearch in this case it works as expected :+1: 
However, it can be discussed if it would be more logical to show it above the query rows? This shouldn't be a blocker, but good to circulate with UX.

![image](https://user-images.githubusercontent.com/1668778/57375011-962ed800-719c-11e9-8884-dd4d6cac736c.png)

Found an interesting thing when testing elasticsearch. I get an error reported in explore which is not query related (no network error), but I cannot see any related information like stacktrace/console output - this feels like a bug.

![image](https://user-images.githubusercontent.com/1668778/57375118-e312ae80-719c-11e9-90a2-e847281b42b8.png)

torkelo commented:
now that we have branched of v6.2 this should be ok to merge

hugohaggmark commented:
@torkelo I'm working on the last error handling stuff, should be ready for merge after lunch

hugohaggmark committed:
Refactor: Makes it possible to return result even when there are errors

hugohaggmark committed:
Fix: Fixes digest issue with Angular editors

hugohaggmark committed:
Refactor: Adds tests for explore utils

hugohaggmark commented:
So this got tricker then I could imaging at first, and I'm really not sure about this solution. 

Today in master we show results for any QueryRow that doesn't have an error for Loki/Prom. To accomplish this in this PR
the Loki/Prom query errors are part of the success result because otherwise we would terminate the whole query batch.

As a plus this PR does fix issues with inconsistent refIds, hinting, interface alignment and digest issues with Angular editors. What do you say about the resulting error handling @marefr @davkal @torkelo ?

I guess if we didn't bother to show results when there are QueryRows with errors the solution would be much cleaner and simpler.

torkelo commented:
I don’t think showing results and errors at the same time  is that important if it complicates the code. Has not been a major source of complaints in how the dashboard panel batches / executes queries.

marefr commented:
Yeah I agree with Torkel, but since it worked that way before I was afraid that it was important to support that. Guess input from @davkal is needed to know for sure.

davkal commented:
During troubleshooting it's desirable to see as much as you can quickly. If results shown are incomplete that's fine as long as there is an indicator that results are incomplete due to a query in a particular row. So not showing any results if one of the rows has an error IMO is a worse UX, but I'd be happy to accept it for now, hoping that we can disentangle this again in the future.

davkal commented:
Seeing odd behavior with Loki: hitting Enter after a entering a query runs 4 queries in parallel (1 row). Hitting the Run button runs 2 queries in parallel.

davkal commented:
Similar thing with Prometheus. Hitting Enter runs 2 sets of queries, the first of which gets cancelled:
![Screenshot 2019-05-09 at 16 52 06](https://user-images.githubusercontent.com/859729/57463342-e5e3d100-727a-11e9-8d35-bbcfc1e46532.png)

hugohaggmark committed:
Refactor: Breakes current behaviour of always returning a result even if Query fails

hugohaggmark commented:
> Similar thing with Prometheus. Hitting Enter runs 2 sets of queries, the first of which gets cancelled:
> ![Screenshot 2019-05-09 at 16 52 06](https://user-images.githubusercontent.com/859729/57463342-e5e3d100-727a-11e9-8d35-bbcfc1e46532.png)

@davka I've looked into this, and this strange behaviour is already in `master` and not a side effect of this PR. I suggest we create a new issue.

hugohaggmark commented:
> During troubleshooting it's desirable to see as much as you can quickly. If results shown are incomplete that's fine as long as there is an indicator that results are incomplete due to a query in a particular row. So not showing any results if one of the rows has an error IMO is a worse UX, but I'd be happy to accept it for now, hoping that we can disentangle this again in the future.

I've reverted my code and introduced this `breaking UX change` in favour of simplifying the code. But if in the future this is a desirable UX we now know what it takes to make it work.

hugohaggmark committed:
Fix: Fixes Prettier error

marefr reviewed (changes requested):
The revert works as expected :+1: 

Still got the elasticsearch processing error. In master, I got no error in Explore UI, but I do on this branch :+1: 
![image](https://user-images.githubusercontent.com/1668778/57514317-7ddfca00-7310-11e9-9d6d-0ceb4e92ac35.png)

In master I get this in the console, but nothing in the console in this branch and this seems like a bug to me:
![image](https://user-images.githubusercontent.com/1668778/57513859-8b488480-730f-11e9-81b2-853eb80db2b1.png)

http://localhost:3000/explore?left=%5B%22now-6h%22,%22now%22,%22gdev-elasticsearch-v6-metrics%22,%7B%22datasource%22:%22gdev-elasticsearch-v6-metrics%22,%22metrics%22:%5B%7B%22type%22:%22raw_document%22,%22id%22:%221%22,%22$$hashKey%22:%22object:135%22,%22field%22:%22select%20field%22,%22settings%22:%7B%22size%22:500%7D,%22meta%22:%7B%7D%7D%5D,%22bucketAggs%22:%5B%5D,%22timeField%22:%22@timestamp%22,%22query%22:%22%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D

hugohaggmark committed:
Fix: Adds back console.log for erroneous querys

marefr commented on public/app/plugins/datasource/prometheus/specs/completer.test.ts:
```diff
@@ -5,6 +5,7 @@ import { DataSourceInstanceSettings } from '@grafana/ui';
 import { PromOptions } from '../types';
 import { TemplateSrv } from 'app/features/templating/template_srv';
 import { TimeSrv } from 'app/features/dashboard/services/TimeSrv';
+import { IQService } from 'angular';
```
What is this?

hugohaggmark replied:
Typings for q$

marefr replied:
Haha okay. I thought there was some service for https://en.wikipedia.org/wiki/Intelligence_quotient :laughing:

marefr commented on packages/grafana-ui/src/utils/moment_wrapper.ts:
```diff
@@ -47,7 +47,7 @@ export interface DateTimeDuration {
 export interface DateTime extends Object {
   add: (amount?: DateTimeInput, unit?: DurationUnit) => DateTime;
-  diff: (b: DateTimeInput, unitOfTime?: DurationInput, precise?: boolean) => number;
+  diff: (amount: DateTimeInput, unit?: DurationUnit, truncate?: boolean) => number;
```
Separate PR?

hugohaggmark replied:
Merge with master gave this

marefr commented on public/app/features/explore/state/actions.ts:
```diff
@@ -423,6 +423,8 @@ export function processQueryErrors(
       return;
     }
 
+    console.log(response); // To help finding problems with query syntax
```
Maybe nitpicking but console.error would be nicer

hugohaggmark replied:
Will fix

marefr reviewed (approved):
Looks good, but left some comments.

hugohaggmark committed:
Refactor: Changes console.log to console.error
